File: python/open_guardrail/guards/input_sanitize.py
"""Sanitizes input by stripping dangerous characters."""

import re
import time
import unicodedata
from typing import List

from open_guardrail.core import GuardResult


class _InputSanitize:
    def __init__(self, *, action: str = "block", strip_html: bool = True, normalize_unicode: bool = True, remove_null_bytes: bool = True, max_newlines: int = 5) -> None:
        self.name = "input-sanitize"
        self.action = action
        self.strip_html = strip_html
        self.normalize = normalize_unicode
        self.remove_null = remove_null_bytes
        self.max_nl = max_newlines

    def check(self, text: str, stage: str = "input") -> GuardResult:
        start = time.perf_counter()
        sanitized = text
        changes: List[str] = []
        if self.remove_null and "\0" in sanitized:
            sanitized = sanitized.replace("\0", "")
            changes.append("null-bytes")
        if self.strip_html and re.search(r"<[^>]+>", sanitized):
            sanitized = re.sub(r"<[^>]*>", "", sanitized)
            changes.append("html-tags")
        if self.normalize:
            normalized = unicodedata.normalize("NFC", sanitized)
            if normalized != sanitized:
                sanitized = normalized
                changes.append("unicode-normalization")
        nl_pat = re.compile(r"\n{" + str(self.max_nl + 1) + r",}")
        if nl_pat.search(sanitized):
            sanitized = nl_pat.sub("\n" * self.max_nl, sanitized)
            changes.append("excessive-newlines")
        modified = len(changes) > 0
        elapsed = (time.perf_counter() - start) * 1000
        if not modified:
            return GuardResult(guard_name="input-sanitize", passed=True, action="allow", latency_ms=round(elapsed, 2))
        if self.action == "mask":
            return GuardResult(guard_name="input-sanitize", passed=True, action="override", override_text=sanitized, latency_ms=round(elapsed, 2), details={"sanitized": changes})
        return GuardResult(guard_name="input-sanitize", passed=False, action=self.action, message=f"Input sanitized: {', '.join(changes)}", latency_ms=round(elapsed, 2), details={"sanitized": changes})


def input_sanitize(*, action: str = "block", strip_html: bool = True, normalize_unicode: bool = True, remove_null_bytes: bool = True, max_newlines: int = 5) -> _InputSanitize:
    return _InputSanitize(action=action, strip_html=strip_html, normalize_unicode=normalize_unicode, remove_null_bytes=remove_null_bytes, max_newlines=max_newlines)

File: packages/guards/src/input-sanitize.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface InputSanitizeOptions {
  action: 'block' | 'warn' | 'mask';
  /** Strip HTML tags (default true) */
  stripHtml?: boolean;
  /** Normalize unicode (default true) */
  normalizeUnicode?: boolean;
  /** Remove null bytes (default true) */
  removeNullBytes?: boolean;
  /** Trim excessive whitespace (default true) */
  trimWhitespace?: boolean;
  /** Max consecutive newlines (default 5) */
  maxNewlines?: number;
}

export function inputSanitize(options: InputSanitizeOptions): Guard {
  const stripHtml = options.stripHtml ?? true;
  const normalizeUnicode = options.normalizeUnicode ?? true;
  const removeNullBytes = options.removeNullBytes ?? true;
  const trimWs = options.trimWhitespace ?? true;
  const maxNl = options.maxNewlines ?? 5;

  return {
    name: 'input-sanitize',
    version: '0.1.0',
    description: 'Sanitizes input by stripping dangerous characters and normalizing text',
    category: 'security',
    supportedStages: ['input'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      let sanitized = text;
      const changes: string[] = [];

      if (removeNullBytes && /\0/.test(sanitized)) {
        sanitized = sanitized.replace(/\0/g, '');
        changes.push('null-bytes');
      }

      if (stripHtml && /<[^>]+>/.test(sanitized)) {
        sanitized = sanitized.replace(/<[^>]*>/g, '');
        changes.push('html-tags');
      }

      if (normalizeUnicode) {
        const normalized = sanitized.normalize('NFC');
        if (normalized !== sanitized) {
          sanitized = normalized;
          changes.push('unicode-normalization');
        }
      }

      if (trimWs) {
        const trimmed = sanitized.replace(/[ \t]+/g, ' ').trim();
        if (trimmed !== sanitized) {
          sanitized = trimmed;
          changes.push('whitespace');
        }
      }

      const nlPattern = new RegExp(`\n{${maxNl + 1},}`, 'g');
      if (nlPattern.test(sanitized)) {
        sanitized = sanitized.replace(nlPattern, '\n'.repeat(maxNl));
        changes.push('excessive-newlines');
      }

      const modified = changes.length > 0;

      if (!modified) {
        return {
          guardName: 'input-sanitize',
          passed: true,
          action: 'allow',
          latencyMs: Math.round(performance.now() - start),
        };
      }

      if (options.action === 'mask') {
        return {
          guardName: 'input-sanitize',
          passed: true,
          action: 'override',
          overrideText: sanitized,
          latencyMs: Math.round(performance.now() - start),
          details: { sanitized: changes },
        };
      }

      return {
        guardName: 'input-sanitize',
        passed: false,
        action: options.action,
        latencyMs: Math.round(performance.now() - start),
        details: { sanitized: changes },
      };
    },
  };
}
